File: maple-engine-plugins/maple-spark3/maple-spark3-data-calc/src/main/scala/org/xi/maple/datacalc/ArgsParser.scala
package org.xi.maple.datacalc

import org.slf4j.{Logger, LoggerFactory}

import scala.annotation.tailrec
import scala.collection.mutable

object ArgsParser {

  private val log: Logger = LoggerFactory.getLogger(ArgsParser.getClass)

  val usage: String =
    """
      |Usage: spark-submit maple.jar [-d <configData> | -f <configFile>]
      |""".stripMargin

  def getParams(args: Array[String]): mutable.Map[String, String] = {
    @tailrec
    def nextOption(map: mutable.Map[String, String], list: List[String]): mutable.Map[String, String] = {
      list match {
        case Nil => map
        case ("-h" | "--help") :: _ =>
          println(usage)
          System.exit(0)
          null
        case ("-f" | "--file") :: value :: tail =>
          nextOption(map ++ Map("file" -> value), tail)
        case ("-d" | "--data") :: value :: tail =>
          nextOption(map ++ Map("data" -> value), tail)
        case key :: value :: tail =>
          if (!key.startsWith("-")) {
            nextOption(map, value :: tail)
          } else if (value.startsWith("-")) {
            nextOption(map ++ Map(key.stripPrefix("-").stripPrefix("-") -> ""), value :: tail)
          } else {
            nextOption(map ++ Map(key.stripPrefix("-").stripPrefix("-") -> value), tail)
          }
        case _ :: tail =>
          nextOption(map, tail)
      }
    }

    val map = mutable.Map[String, String]()
    val result = nextOption(map, args.toList)

    if (result.contains("file") && result.contains("data")) {
      log.error("Config file or config data string cannot be set up at the same time")
      System.exit(1)
    }
    if (!result.contains("file") && !result.contains("data")) {
      log.error("Config file or config data string should be set up")
      System.exit(1)
    }
    result
  }
}

File: maple-engine-plugins/maple-flink/apps/maple-flink-data-calc/src/main/java/org/xi/maple/datacalc/flink/ArgsParser.java
package org.xi.maple.datacalc.flink;

import org.apache.commons.cli.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class ArgsParser {
    private static final Logger logger = LoggerFactory.getLogger(ArgsParser.class);

    private static final Options options = getOptions();

    public static CommandLine getCommandLine(String[] args) {

        DefaultParser parser = new DefaultParser();
        CommandLine cmd = null;
        try {
            cmd = parser.parse(options, args);
            if (cmd.hasOption("h")) {
                printUsage();
                System.exit(0);
            }
            if (cmd.hasOption("f") && cmd.hasOption("d")) {
                logger.error("Config file or config data string cannot be set up at the same time");
                System.exit(1);
            }
            if (!cmd.hasOption("f") && !cmd.hasOption("d")) {
                logger.error("Config file or config data string should be set up");
                System.exit(1);
            }
        } catch (ParseException e) {
            System.err.println(e.getMessage());
            System.exit(1);
        }
        return cmd;
    }

    private static Options getOptions() {
        Options options = new Options();
        options.addOption("h", "help", false, "显示帮助信息");
        options.addOption("f", "file", true, "配置地址");
        options.addOption("d", "data", true, "配置内容");
        options.addOption("m", "mode", true, "执行模式");
        return options;
    }

    private static void printUsage() {
        HelpFormatter hf = new HelpFormatter();
        hf.printHelp("Usage", options);
    }
}
